DKImageManager/Data/DKGroupDataManager.swift

```diff
@@ -39,35 +39,50 @@ public class DKGroupDataManager: DKBaseManager, PHPhotoLibraryChangeObserver {
 		
 		PHPhotoLibrary.shared().unregisterChangeObserver(self)
 	}
-	
-	public func fetchGroups(_ completeBlock: (_ groups: [String]?, _ error: NSError?) -> Void) {
+
+	public func fetchGroups(_ completeBlock: @escaping (_ groups: [String]?, _ error: NSError?) -> Void) {
 		if let assetGroupTypes = self.assetGroupTypes {
-			if self.groups != nil {
-				completeBlock(self.groupIds, nil)
-				return
-			}
-			
-			var groups: [String : DKAssetGroup] = [:]
-			var groupIds: [String] = []
-			
-			for (_, groupType) in assetGroupTypes.enumerated() {
-                let fetchResult = PHAssetCollection.fetchAssetCollections(with: self.collectionTypeForSubtype(groupType),
-				                                                                  subtype: groupType,
-				                                                                  options: nil)
-                fetchResult.enumerateObjects({ (collection, index, stop) in
-                    let assetGroup = DKAssetGroup()
-                    assetGroup.groupId = collection.localIdentifier
-                    self.updateGroup(assetGroup, collection: collection)
-                    if self.showsEmptyAlbums || assetGroup.totalCount > 0 {
-                        groups[assetGroup.groupId] = assetGroup
-                        groupIds.append(assetGroup.groupId)
-                    }
-                })
+			DispatchQueue.global(qos: .userInteractive).async {
+				[weak self] in
+				guard let strongSelf = self else {
+					return
+				}
+
+				guard strongSelf.groups == nil else {
+					DispatchQueue.main.async {
+						completeBlock(strongSelf.groupIds, nil)
+					}
+					return
+				}
+
+				var groups: [String : DKAssetGroup] = [:]
+				var groupIds: [String] = []
+
+				for (_, groupType) in assetGroupTypes.enumerated() {
+					let fetchResult = PHAssetCollection.fetchAssetCollections(with: strongSelf.collectionTypeForSubtype(groupType),
+							subtype: groupType,
+							options: nil)
+					fetchResult.enumerateObjects({ (collection, index, stop) in
+						let assetGroup = DKAssetGroup()
+						assetGroup.groupId = collection.localIdentifier
+						strongSelf.updateGroup(assetGroup, collection: collection)
+						if strongSelf.showsEmptyAlbums || assetGroup.totalCount > 0 {
+							groups[assetGroup.groupId] = assetGroup
+							groupIds.append(assetGroup.groupId)
+						}
+						strongSelf.updatePartial(groups: groups, groupIds: groupIds, completeBlock: completeBlock)
```

Comment:
It is wired, we re-assign `self.groups` and `self.groupIds` maybe we should just update them, not re-assign them, and it seems we will invoke callback several times here.